cmd/tools/rpcwrappers/main.go

```diff
@@ -0,0 +1,500 @@
+	"io"
+	"os"
+	"reflect"
+	"strings"
+	"text/template"
+
+	"go.temporal.io/api/workflowservice/v1"
+	"golang.org/x/exp/slices"
+
+	"go.temporal.io/server/api/adminservice/v1"
+	"go.temporal.io/server/api/historyservice/v1"
+	"go.temporal.io/server/api/matchingservice/v1"
+)
+
+type (
+	service struct {
+		name               string
+		clientType         reflect.Type
+		clientGenerator    func(io.Writer, service)
+		metricGenerator    func(io.Writer, service)
+		retryableGenerator func(io.Writer, service)
+	}
+)
+
+var (
+	services = []service{
+		service{
+			name:               "frontend",
+			clientType:         reflect.TypeOf((*workflowservice.WorkflowServiceClient)(nil)),
+			clientGenerator:    generateFrontendOrAdminClient,
+			metricGenerator:    generateFrontendOrAdminMetricClient,
+			retryableGenerator: generateRetryableClient,
+		},
+		service{
+			name:               "admin",
+			clientType:         reflect.TypeOf((*adminservice.AdminServiceClient)(nil)),
+			clientGenerator:    generateFrontendOrAdminClient,
+			metricGenerator:    generateFrontendOrAdminMetricClient,
+			retryableGenerator: generateRetryableClient,
+		},
+		service{
+			name:               "history",
+			clientType:         reflect.TypeOf((*historyservice.HistoryServiceClient)(nil)),
+			clientGenerator:    generateHistoryClient,
+			metricGenerator:    generateHistoryOrMatchingMetricClient,
+			retryableGenerator: generateRetryableClient,
+		},
+		service{
+			name:               "matching",
+			clientType:         reflect.TypeOf((*matchingservice.MatchingServiceClient)(nil)),
+			clientGenerator:    generateMatchingClient,
+			metricGenerator:    generateHistoryOrMatchingMetricClient,
+			retryableGenerator: generateRetryableClient,
+		},
+	}
+
+	longPollContext = map[string]bool{
+		"client.frontend.ListArchivedWorkflowExecutions": true,
+		"client.frontend.PollActivityTaskQueue":          true,
+		"client.frontend.PollWorkflowTaskQueue":          true,
+	}
+	largeTimeoutContext = map[string]bool{
+		"client.admin.GetReplicationMessages": true,
+	}
+	ignoreMethod = map[string]bool{
+		// these are non-standard implementations. do not generate.
+		"client.history.DescribeHistoryHost":    true,
+		"client.history.GetReplicationMessages": true,
+		"client.history.GetReplicationStatus":   true,
+		// these need to pick a partition. too complicated.
+		"client.matching.AddActivityTask":       true,
+		"client.matching.AddWorkflowTask":       true,
+		"client.matching.PollActivityTaskQueue": true,
+		"client.matching.PollWorkflowTaskQueue": true,
+		"client.matching.QueryWorkflow":         true,
+		// these do forwarding stats. too complicated.
+		"metricsClient.matching.AddActivityTask":       true,
+		"metricsClient.matching.AddWorkflowTask":       true,
+		"metricsClient.matching.PollActivityTaskQueue": true,
+		"metricsClient.matching.PollWorkflowTaskQueue": true,
+		"metricsClient.matching.QueryWorkflow":         true,
+	}
+)
+
+func writeCopyright(w io.Writer) {
```

Comment:
We have a licensegen.go. is it possible to reuse it?

Reply:
sure, I made it read the root LICENSE file similar to that